Teach bootstrap about target files vs target triples

shepmaster opened the pull request:
`rustc` allows passing in predefined target triples as well as JSON
target specification files. This change allows bootstrap to have the
first inkling about those differences. This allows building a
cross-compiler for an out-of-tree architecture (even though that
compiler won't work for other reasons).

Even if no one ever uses this functionality, I think the newtype
around the `Interned<String>` improves the readability of the code.

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #74245) made this pull request unmergeable. Please resolve the merge conflicts.

Mark-Simulacrum commented on src/bootstrap/config.rs:
```diff
@@ -158,6 +159,67 @@ pub struct Config {
+                .file_stem()
+                .expect("Target specification file has no file stem")
+                .to_str()
+                .expect("Target specification file is not UTF-8");
```
```suggestion
                .expect("Target specification file stem is not UTF-8");
```

Mark-Simulacrum reviewed (commented):
r=me with nit fixed, and presuming this works for your use case :)

shepmaster commented:
@bors r=@Mark-Simulacrum rollup

bors commented:
:pushpin: Commit c24b96dcb58a4f00d055128ea5caabd62f456b7f has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum c24b96dcb58a4f00d055128ea5caabd62f456b7f -->
<!-- homu: {"type":"Approved","sha":"c24b96dcb58a4f00d055128ea5caabd62f456b7f","approver":"Mark-Simulacrum"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 20, this pull request will be tested once the tree is reopened

Manishearth commented:
@bors r- rollup-

This broke the `ignore-wasm32-bare` annotation in compiletest

https://github.com/rust-lang/rust/pull/74308#issuecomment-657894134

Also I'd avoid marking build system PRs as `rollup`, that annotation is for things which have a very low chance of failing CI (doc PRs, etc)

eddyb commented:
@Manishearth I'm no so sure this PR broke wasm, the test fails w/o it: https://github.com/rust-lang/rust/pull/74309#issuecomment-657917670

shepmaster commented:
Based on previous comments that this doesn't have an issue but leaving off rollup as directed...

@bors r=@Mark-Simulacrum

bors commented:
:pushpin: Commit c24b96dcb58a4f00d055128ea5caabd62f456b7f has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum c24b96dcb58a4f00d055128ea5caabd62f456b7f -->
<!-- homu: {"type":"Approved","sha":"c24b96dcb58a4f00d055128ea5caabd62f456b7f","approver":"Mark-Simulacrum"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #74375) made this pull request unmergeable. Please resolve the merge conflicts.

shepmaster committed:
Teach bootstrap about target files vs target triples

`rustc` allows passing in predefined target triples as well as JSON
target specification files. This change allows bootstrap to have the
first inkling about those differences. This allows building a
cross-compiler for an out-of-tree architecture (even though that
compiler won't work for other reasons).

Even if no one ever uses this functionality, I think the newtype
around the `Interned<String>` improves the readability of the code.

shepmaster commented:
Let's see what causes a conflict this time...

@bors r=@Mark-Simulacrum

bors commented:
:pushpin: Commit e2b337dc57b5309d19e25811d3de340cc7567285 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum e2b337dc57b5309d19e25811d3de340cc7567285 -->
<!-- homu: {"type":"Approved","sha":"e2b337dc57b5309d19e25811d3de340cc7567285","approver":"Mark-Simulacrum"} -->
